compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,272 @@
+                                      LayoutProvider *provider) {
+  // If this is a result of a "to_simd" op, use the source value of it.
+  if (auto toSIMD = value.getDefiningOp<IREE::VectorExt::ToSIMDOp>()) {
+    value = cast<TypedValue<VectorType>>(toSIMD.getInput());
+    return value;
+  }
+  // Create a "to_simt" op to convert the value to the distributed layout.
+  SmallVector<int64_t> distributedShape = provider->getDistributedShape(value);
+  VectorType distributedType =
+      VectorType::get(distributedShape, value.getType().getElementType());
+  auto toSIMT = rewriter.create<IREE::VectorExt::ToSIMTOp>(
+      value.getLoc(), distributedType, value);
+  return toSIMT.getResult();
+}
+
+void replaceOpWithDistributedValues(RewriterBase &rewriter, Operation *op,
+                                    LayoutProvider *provider,
+                                    ValueRange values) {
+  // Replace all OpResults with the given values.
+  SmallVector<Value> replacements;
+  for (OpResult opResult : op->getOpResults()) {
+    Value replacement = values[opResult.getResultNumber()];
+    // If this value is a vector type, it must be converted back to simd.
+    if (isa<VectorType>(replacement.getType())) {
+      auto oldResult = cast<TypedValue<VectorType>>(opResult);
+      // Create a toSIMD op to convert the value back to the simd.
+      rewriter.setInsertionPointAfterValue(oldResult);
+      auto toSIMD = rewriter.create<IREE::VectorExt::ToSIMDOp>(
+          oldResult.getLoc(), oldResult.getType(), replacement);
+      // Clone the layout to the new value.
+      provider->getAnalysis().cloneLayoutInformationToNewValue(
+          oldResult, toSIMD.getResult());
+      // Add to replacements.
+      replacement = toSIMD.getResult();
+    }
+    replacements.push_back(replacement);
+  }
+
+  rewriter.replaceOp(op, replacements);
+}
+
+class DistributionRewriter : public IRRewriter, public RewriterBase::Listener {
+public:
+  DistributionRewriter(MLIRContext *ctx, DenseSet<Operation *> &erasedOps,
+                       SmallVector<Operation *> &worklist,
+                       VectorLayoutAnalysis &analysis)
+      : IRRewriter(ctx), erasedOps(erasedOps), worklist(worklist),
+        analysis(analysis) {
+    setListener(this);
+  }
+
+protected:
+  void notifyOperationRemoved(Operation *op) override { erasedOps.insert(op); }
+
+  void notifyOperationInserted(Operation *op) override {
+    // For now, do nothing.
+    // TODO: Check if the operation can still be distributed and try to
+    // distribute it. Not sure how to check this? Maybe we can check if the
+    // analysis knows about this value and assume that this value needs to be
+    // distributed. This is needed for SIMD -> SIMD rewrites.
+  }
+
+private:
+  // A reference to the set of operations that have been erased.
+  DenseSet<Operation *> &erasedOps;
+  // A reference to the worklist of operations that need to be distributed.
+  SmallVector<Operation *> &worklist;
```

Comment:
we can let the caller decide what size to use
```suggestion
  // A reference to the worklist of operations that need to be distributed.
  SmallVectorImpl<Operation *> &worklist;
```